### PR TITLE
Reimplement "Mask Scoring R-CNN"

### PR DESCRIPTION
**Pull request by yhcao6**
The code refers to https://github.com/zjhuang22/maskscoring_rcnn

|           | model         | mask AP | box AP |
|-----------|---------------|---------|--------|
| official  | Mask RCNN R50 | 34.2    | 37.8   |
| official  | MS RCNN R50   | 35.6    | 37.9   |
| this repo | Mask RCNN R50 | 34.3    | 37.4   |
| this repo | MS RCNN R50   | 35.6    | 37.5   |

### CODE REVIEW

File: mmdet/core/evaluation/coco_utils.py [hunk 1]
```diff
@@ -97,7 +101,7 @@ def proposal2json(dataset, results):
             data['score'] = float(bboxes[i][4])
             data['category_id'] = 1
             json_results.append(data)
-    return json_results
+    return dict(proposal=json_results)
```
Comment (hellock) [inline mmdet/core/evaluation/coco_utils.py]: The return types of `xxx2json` should remain the same.

File: mmdet/core/mask/mask_target.py [hunk 2]
```diff
@@ -34,3 +34,40 @@ def mask_target_single(pos_proposals, pos_assigned_gt_inds, gt_masks, cfg):
...
     return mask_targets
+
+
+def mask_iou_target(pos_proposals_list, pos_assigned_gt_inds_list,
+                    gt_masks_list, mask_preds, mask_targets):
+    area_ratios = map(mask_iou_target_single, pos_proposals_list,
+                      pos_assigned_gt_inds_list, gt_masks_list)
+    area_ratios = torch.cat(list(area_ratios))
+    assert mask_targets.size(0) == area_ratios.size(0)
+    mask_pred = (mask_preds > 0.5).float()  # binarize mask pred
```
Comment (hellock) [inline mmdet/core/mask/mask_target.py]: The threshold should be set in the config.

File: configs/ms_rcnn/ms_rcnn_r101_caffe_fpn_1x.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
+# model settings
+model = dict(
+    type='MaskRCNN',
```
Comment (hellock) [inline configs/ms_rcnn/ms_rcnn_r101_caffe_fpn_1x.py]: Add a new detector `MaskScoringRCNN`.

File: mmdet/models/detectors/base.py [hunk 4]
```diff
@@ -33,6 +33,11 @@ def with_bbox(self):
     def with_mask(self):
         return hasattr(self, 'mask_head') and self.mask_head is not None
 
+    @property
+    def with_mask_iou(self):
```
Comment (hellock) [inline mmdet/models/detectors/base.py]: Remove this property from the base detector.

File: mmdet/models/detectors/two_stage.py [hunk 5]
```diff
@@ -21,6 +21,7 @@ def __init__(self,
                  bbox_head=None,
                  mask_roi_extractor=None,
                  mask_head=None,
+                 mask_iou_head=None,
```
Comment (hellock) [inline mmdet/models/detectors/two_stage.py]: Move this argument to `MaskScoringRCNN`.

File: mmdet/models/mask_heads/maskiou_head.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import kaiming_init, normal_init
+
+from mmdet.core import mask_iou_target
+from ..registry import HEADS
+
+
+@HEADS.register_module
+class MaskIoUHead(nn.Module):
```
Comment (hellock) [inline mmdet/models/mask_heads/maskiou_head.py]: Add a docstring here.

File: mmdet/core/mask/mask_target.py [hunk 7]
```diff
@@ -34,3 +34,40 @@ def mask_target_single(pos_proposals, pos_assigned_gt_inds, gt_masks, cfg):
...
+    assert mask_targets.size(0) == area_ratios.size(0)
+    mask_pred = (mask_preds > cfg.mask_thr_binary).float()
+    mask_overlaps = (mask_pred * mask_targets).sum((-1, -2))
+    full_areas = mask_targets.sum((-1, -2)) / area_ratios
+    mask_unions = mask_pred.sum((-1, -2)) + full_areas - mask_overlaps
+    mask_iou_targets = mask_overlaps / mask_unions
+    return mask_iou_targets
+
+
+def mask_iou_target_single(pos_proposals, pos_assigned_gt_inds, gt_masks):
```
Comment (hellock) [inline mmdet/core/mask/mask_target.py]: Docstring and comments.

File: mmdet/models/detectors/mask_scoring_rcnn.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
+import torch
+
+from mmdet.core import bbox2roi, bbox2result, build_assigner, build_sampler
+from .two_stage import TwoStageDetector
+from .. import builder
+from ..registry import DETECTORS
+
+
+@DETECTORS.register_module
+class MaskScoringRCNN(TwoStageDetector):
```
Comment (hellock) [inline mmdet/models/detectors/mask_scoring_rcnn.py]: Docstring here.

File: mmdet/models/detectors/test_mixins.py [hunk 9]
```diff
@@ -116,7 +118,7 @@ def simple_test_mask(self,
             segm_result = self.mask_head.get_seg_masks(
                 mask_pred, _bboxes, det_labels, self.test_cfg.rcnn, ori_shape,
                 scale_factor, rescale)
-        return segm_result
+        return segm_result, mask_feats, mask_pred
```
Comment (hellock) [inline mmdet/models/detectors/test_mixins.py]: This change may be moved to `MaskScoringRCNN.simple_test_mask()`.

**Comment by zeenolife** [comment]
I might be wrong here, but shouldn't mask scoring generalize to multi-stage detectors as well?
